### ISSUE TITLE
Build and publish Nimbus Rust code + headers as an XCFramework

### ISSUE THREAD

**Issue by data-sync-user** [mozilla/application-services#4305]
Each application-services release should build and upload a new binary artifact named `MozillaRustComponents.xcframework.zip`. It will contain an iOS “XCFramework” of the compiled Rust code for Nimbus, Viaduct and RustLog supporting both device and simulator builds.

There is a prototype of building this artifact here: [https://github.com/mozilla/application-services/pull/4225](https://github.com/mozilla/application-services/pull/4225|smart-link) 

We need to make it production-ready and incorporate it into our automation.

Broadly, this will involve:

* Defining a new “megazord” crate that includes Nimbus, Viaduct and RustLog.
* Adding a script to build it into an XCFramework bundle.
  * Each framework in the XCFramework bundle should contains:
    * The compiled Rust code as a static lib
    * The corresponding headers
    * An umbrella header and corresponding module map
  * There is a prototype script that just copies files into the expected location [here](https://github.com/mozilla/application-services/pull/4225/files#diff-3729d0e227487c58350ead0b2d76baa018fcca7439fb97e5ec1c2b4620fd1ee1)
  * The prototype currently duplicates some headers; the production version should copy them from their canonical location or generate them on demand.
* Adding a CI task to run this script on release and upload the resulting artifact.
  * The existing scripts for publishing our Carthage artifact will be a useful start here.

┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/SYNC-2674)
┆Story Points: 8
┆Epic: Unified a-s swift packages for iOS
┆Sprint End Date: 2021-09-03

**Comment by data-sync-user** [comment]
➤ Ryan Kelly commented:

A brief update here: I’ve landed a few preparatory commits in support of this ticket, and have updated the PR to try building the XCFramework in CI:

  * https://github.com/mozilla/application-services/pull/4225 ( https://github.com/mozilla/application-services/pull/4225|smart-link ) 

Once that’s working we’ll be able to try out the above Acceptance Criteria with a built artifact, and then hopefully the “do it automatically on each release” part will be straightforward.

**Comment by data-sync-user** [comment]
➤ Ryan Kelly commented:

The XCFramework artifact appears to be building successfully in CI. Over in https://mozilla-hub.atlassian.net/browse/SYNC-2675 ( https://mozilla-hub.atlassian.net/browse/SYNC-2675|smart-link ) I posted some notes about using the resulting artifact in a build and AFAICT it’s working correctly. It needs to be tested out on an M1 though.

The PR itself is blocked on some dependency updates (specifically, updating UniFFI to the latest release) but is otherwise ready for review, and I walked through it with Sammy Khamis and Tarik Eshaq earlier today.
